src/core/seat/input-method-relay.cpp

```diff
@@ -60,11 +60,20 @@ wf::input_method_relay::input_method_relay()
         // We ignore such commit requests so it doesn't have any affect on the
         // new window. Even when the previous window isn't preediting when
         // switching focus, it doesn't have any bad effect to the new window anyway.
-        if (focus_just_changed)
+        //
+        // Preediting can be disabled (selectively or globally) because
+        // application bugs were observed. In this case, we see only commits
+        // but no preedit strings, so we need care the timing.
+        static std::chrono::milliseconds focus_change_duration{100};
+        if (last_focus_changed.has_value())
```

Comment:
I looked at the protocol and I believe what we are doing here is not a real fix for the issue. The protocol includes serials which is what we really need here. So here's the deal as far as I understand it:

- When we switch focus or do something of the sort, we can send a `done` event. This generates a new serial number, serial numbers are counted from 0 to infinity and are incremented for each done event. I briefly looked through wlroots implementation but it might be wrong though (because they reset input_method->current_serial on each commit, which might actually have outdated information!). So we might need to keep track of it ourselves.

- For each commit from the text input, we receive the last serial that the input method has received so far (wlroots stores that in input_method->current_serial before emitting the commit event). So this means: until the input method has actually received our new configuration, it will send a commit with an outdated serial.

Reply:
No, I tried to add a `done` event after `set_focus`. It sends a commit with a correct serial. It doesn't make any difference whether we've changed focus. The input method receives a sequence of `done` events and counts it correctly.

Reply:
> No, I tried to add a `done` event after `set_focus`. It sends a commit with a correct serial. It doesn't make any difference whether we've changed focus. The input method receives a sequence of `done` events and counts it correctly.

Did you also try to check whether the serial belongs to the current focus or is an older event in the commit handler?

Reply:
Actually there is already a `done` event when switching focus, and so by adding another `done` event, the input method replies with two `commit`s, with the same serial.

What do you mean by `the serial belongs to the current focus`? The serial is just a counter of `done` events. (I'm reading `WAYLAND_DEBUG` here.)